window/src/os/wayland/drag_and_drop.rs

```diff
@@ -0,0 +1,116 @@
+use crate::connection::ConnectionOps;
+use crate::wayland::{read_pipe_with_timeout, WaylandConnection};
+use filedescriptor::Pipe;
+use smithay_client_toolkit as toolkit;
+use std::os::unix::io::AsRawFd;
+use toolkit::reexports::client::protocol::wl_data_device::Event as DataDeviceEvent;
+use toolkit::reexports::client::protocol::wl_data_offer::WlDataOffer;
+use url::Url;
+use wayland_client::protocol::wl_data_device_manager::DndAction;
+
+#[derive(Default)]
+pub struct DragAndDrop {
+    offer: Option<(u32, WlDataOffer)>, // surface id, offer
+}
+
+pub const URI_MIME_TYPE: &str = "text/uri-list";
+
+impl DragAndDrop {
+    pub fn handle_data_event(&mut self, event: DataDeviceEvent) {
+        match event {
+            DataDeviceEvent::Enter {
+                serial,
+                surface,
+                id,
+                ..
+            } => {
+                if let Some(offer) = id {
+                    offer.accept(serial, Some(URI_MIME_TYPE.to_string()));
+                    offer.set_actions(
+                        DndAction::None | DndAction::Copy | DndAction::Move,
+                        DndAction::None,
+                    );
+                    self.offer = Some((surface.as_ref().id(), offer));
+                }
+            }
+            DataDeviceEvent::Leave => {
+                if let Some((_, offer)) = &self.offer {
+                    offer.destroy();
+                    self.offer = None;
+                }
+            }
+            DataDeviceEvent::Motion { .. } => {}
+            DataDeviceEvent::Drop => {
+                let res = if let Some((surface_id, offer)) = &self.offer {
+                    if let Ok(pipe) =
+                        Pipe::new().map_err(|err| log::error!("Unable to create pipe: {:#}", err))
+                    {
+                        offer.receive(URI_MIME_TYPE.to_string(), pipe.write.as_raw_fd());
+                        let read = pipe.read;
+                        offer.finish();
+                        let copied_surface_id = *surface_id;
+                        self.offer = None;
+                        Some((copied_surface_id, read))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                };
```

Comment:
It feels a little awkward to read with the different branches returning `None`. I would suggest re-phrasing this using a helper function and taking advantage of the `?` operator, something like:

```rust
struct SurfaceAndPipe {
   surface_id: u32,
   read: FileDescriptor,
}

/// Takes the current offer, if any, and initiates a receive into a pipe,
/// returning that surface and pipe descriptor.
fn create_pipe_for_drop(&mut self) -> Option<SurfaceAndPipe> {
  let offer = self.offer.take()?;
  let pipe = Pipe::new().map_err(|err| log::error!("Unable to create pipe: {:#}", err)).ok()?;
  offer.receive(URI_MIME_TYPE.to_string(), pipe.write.as_raw_fd());
  let read = pipe.read;
  offer.finish();
  Some(SurfaceAndPipe {
    surface_id: offer.surface_id,
    read
  })
}
```

Then this block might end up looking something like this:

```rust
                if let Some(res) = self.create_pipe_for_drop() {
                    std::thread::spawn(move || {
                        read_uris_from_pipe(res, window_id)
                    });
                 }
```